apiserver/internal/handlers/resource/validate/validate.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package validate
+
+import (
+	"crypto/sha512"
+	"encoding/hex"
+	"io"
+	"os"
+	"syscall"
+
+	"github.com/juju/juju/core/logger"
+	"github.com/juju/juju/internal/errors"
+)
+
+const (
+	// ErrUnexpectedHash is returned when the hash of the downloaded resource
+	// does not match the expected hash.
+	ErrUnexpectedHash = errors.ConstError("downloaded resource has unexpected hash")
+
+	// ErrUnexpectedSize is returned when the size of the downloaded resources
+	// does not match the expected size.
+	ErrUnexpectedSize = errors.ConstError("downloaded resource has unexpected size (this could be because the controller disk does not have enough space)")
+)
+
+// FileSystem defines a file system for modifying files on a users system.
+type FileSystem interface {
+	// CreateTemp creates a new temporary file in the directory dir,
+	// opens the file for reading and writing, and returns the resulting file.
+	CreateTemp(dir, pattern string) (*os.File, error)
+
+	// Open opens the named file for reading. If successful, methods on
+	// the returned file can be used for reading; the associated file
+	Open(name string) (*os.File, error)
+
+	// Remove removes the named file or (empty) directory.
+	Remove(name string) error
+
+	// Statfs returns information about the file system.
+	Statfs(path string, stat *syscall.Statfs_t) error
+}
+
+// Validator validates a resource blob.
+type Validator struct {
+	logger     logger.Logger
+	fileSystem FileSystem
+}
+
+// NewValidator returns a new resource blob validator.
+func NewValidator(logger logger.Logger, fileSystem FileSystem) *Validator {
+	return &Validator{
+		logger:     logger,
+		fileSystem: fileSystem,
+	}
+}
+
+// Validate takes a ReadCloser containing a resource blob and checks that
+// the size and hash match the expected values. It downloads the blob to a
+// temporary file and returns a ReadCloser that deletes the temporary file on
+// closure.
+//
+// Returns [ErrUnexpectedHash] if the hash of the downloaded resource does not
+// match the expected hash.
+// Returns [ErrUnexpectedSize] if the size of the downloaded resource does not
+// match the expected size.
+func (v *Validator) Validate(
```

Comment:
This is doing more than validating, this is reading a file of the request body, creating a temp file to push it into.